File: src/main/java/no/fintlabs/model/configuration/CollectionMapping.java
package no.fintlabs.model.configuration;

import lombok.Builder;
import lombok.EqualsAndHashCode;
import lombok.Getter;
import lombok.extern.jackson.Jacksonized;

import java.util.ArrayList;
import java.util.Collection;
import java.util.Optional;

@Getter
@Builder
@EqualsAndHashCode
@Jacksonized
public class CollectionMapping<T> {

    public CollectionMapping(
            Collection<T> elementMappings,
            Collection<FromCollectionMapping<T>> fromCollectionMappings
    ) {
        this.elementMappings = Optional.ofNullable(elementMappings).orElse(new ArrayList<>());
        this.fromCollectionMappings = Optional.ofNullable(fromCollectionMappings).orElse(new ArrayList<>());
    }

    private final Collection<T> elementMappings;

    private final Collection<FromCollectionMapping<T>> fromCollectionMappings;

    @Override
    public String toString() {
        return "Sensitive data omitted";
    }

}

File: src/main/kotlin/no/novari/flyt/mapping/model/configuration/CollectionMapping.kt
package no.novari.flyt.mapping.model.configuration

data class CollectionMapping<T>(
    val elementMappings: Collection<T> = emptyList(),
    val fromCollectionMappings: Collection<FromCollectionMapping<T>> = emptyList(),
) {
    class Builder<T> {
        private var elementMappings: Collection<T>? = null
        private var fromCollectionMappings: Collection<FromCollectionMapping<T>>? = null

        fun elementMappings(elementMappings: Collection<T>) = apply { this.elementMappings = elementMappings }

        fun fromCollectionMappings(fromCollectionMappings: Collection<FromCollectionMapping<T>>) =
            apply { this.fromCollectionMappings = fromCollectionMappings }

        fun build(): CollectionMapping<T> {
            return CollectionMapping(
                elementMappings = elementMappings ?: emptyList(),
                fromCollectionMappings = fromCollectionMappings ?: emptyList(),
            )
        }
    }

    companion object {
        @JvmStatic
        fun <T> builder(): Builder<T> = Builder()
    }
}
